File: design/DataStreamasDisjointIntervals.java
class SummaryRanges {
    TreeSet<int[]> set;

    public SummaryRanges() {
        set = new TreeSet<>((a, b) -> a[0] == b[0] ? a[1] - b[1] : a[0] - b[0]);
    }

    public void addNum(int value) {
        int[] cur = new int[] { value, value };

        if (set.contains(cur))
            return;

        int[] low = set.lower(cur);
        int[] high = set.higher(cur);

        if (high != null && high[0] == value)
            return;

        if (low != null && low[1] + 1 == value && high != null && value + 1 == high[0]) {
            low[1] = high[1];
            set.remove(high);
        }

        else if (low != null && low[1] + 1 >= value) {
            low[1] = Math.max(low[1], value);
        }

        else if (high != null && high[0] - 1 == value) {
            high[0] = value;
        } else {
            set.add(cur);
        }
    }

    public int[][] getIntervals() {
        List<int[]> res = new ArrayList<>();
        for (int[] c : set) {
            res.add(c);
        }
        return res.toArray(new int[res.size()][]);
    }
}

/**
 * Your SummaryRanges object will be instantiated and called as such:
 * SummaryRanges obj = new SummaryRanges();
 * obj.addNum(value);
 * int[][] param_2 = obj.getIntervals();
 */

File: design/DataStreamasDisjointIntervals.py
class SummaryRanges:

    def __init__(self):
        self.nums = [0] * 10002
        self.min = inf
        self.max = -inf

    def addNum(self, value: int) -> None:
        self.nums[value] = 1
        self.min = min(value, self.min)
        self.max = max(value, self.max)

    def getIntervals(self) -> List[List[int]]:
        ans = []
        start = -1
        for i in range(self.min, self.max+2):
            if self.nums[i] > 0:
                if start < 0:
                    start = i
            else:
                if start > -1:
                    ans.append([start, i-1])
                    start = -1
        return ans


# Your SummaryRanges object will be instantiated and called as such:
# obj = SummaryRanges()
# obj.addNum(value)
# param_2 = obj.getIntervals()
